bindings/go/blob/locked_reader.go

```diff
@@ -0,0 +1,62 @@
+package blob
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log/slog"
+	"sync"
+)
+
+// LockedReader provides thread-safe access to blob data by wrapping reads
+// with mutex synchronization. It uses a pipe to stream data while holding
+// a read lock on the underlying blob.
+type LockedReader struct {
+	pr *io.PipeReader
+}
+
+// NewLockedReader creates a thread-safe reader for a blob that acquires a read lock
+// before streaming data. The lock is held during the entire copy operation to ensure
+// consistent access and prevent concurrent writes.
+func NewLockedReader(ctx context.Context, mu *sync.RWMutex, blob ReadOnlyBlob) (io.ReadCloser, error) {
+	// Create pipe for streaming
+
+	var rc io.ReadCloser
+	var err error
+	if rc, err = blob.ReadCloser(); err != nil {
+		return nil, fmt.Errorf("unable to get reader: %w", err)
+	}
+	pr, pw := io.Pipe()
+	go func() {
+		mu.RLock()
+
+		defer func() {
+			if err := rc.Close(); err != nil {
+				slog.ErrorContext(ctx, "unable to close reader", slog.String("error", err.Error()))
+			}
+			mu.RUnlock()
+		}()
+
+		// Copy data through pipe
+		if _, err := io.Copy(pw, rc); err != nil {
+			if err := pw.CloseWithError(fmt.Errorf("unable to copy data: %w", err)); err != nil {
+				slog.ErrorContext(ctx, "unable to close pipe with error", slog.String("error", err.Error()))
+			}
+			return
+		}
+
+		if err := pw.Close(); err != nil {
+			slog.ErrorContext(ctx, "unable to close reader", slog.String("error", err.Error()))
+		}
+	}()
+
+	return &LockedReader{pr: pr}, nil
+}
+
+func (lr *LockedReader) Read(p []byte) (n int, err error) {
+	return lr.pr.Read(p)
```

Comment:
I'm stupid, but shouldn't the Read and the Close use the lock instead of the io.Copy in the go routine?